app/services/paypal_service/api/accounts.rb

```diff
@@ -0,0 +1,225 @@
+    # , success_url: "http://marketplace.com/success
+    # , cancel_url: "http://marketplace.com/cancel
+    # }
+    #
+
+    def billing_agreement_request(community_id, person_id, body)
+      with_billing_agreement_request(body[:description], body[:success_url], body[:cancel_url]) { |billing_agreement_request|
+        account = PaypalAccountStore.update({
+                                              community_id: community_id,
+                                              person_id: person_id,
+                                              billing_agreement_paypal_username_to: billing_agreement_request[:username_to],
+                                              billing_agreement_request_token: billing_agreement_request[:token]
+                                            })
+
+        Result::Success.new(DataTypes.create_billing_agreement_request({
+                                                                         redirect_url: billing_agreement_request[:redirect_url]
+                                                                       }))
+      }
+    end
+
+    ## POST /accounts/:community_id/:person_id/billing_agreement/create?token=EC-123215122362
+    #
+    # Empty body
+    #
+    # Errors:
+    #
+    # - :billing_agreement_not_accepted
+    # - :wrong_account
+
+    def billing_agreement_create(community_id, person_id, request_token)
+      paypal_account = PaypalAccountStore.get(person_id, community_id)
+
+      with_billing_agreement(request_token) { |billing_agreement|
+        with_express_checkout_details(request_token) { |express_checkout_details|
+          if !express_checkout_details[:billing_agreement_accepted]
+            Result::Error.new(:billing_agreement_not_accepted)
+          elsif express_checkout_details[:payer_id] != paypal_account[:payer_id]
+            Result::Error.new(:wrong_account)
+          else
+            account = PaypalAccountStore.update(
+              {
+                community_id: community_id,
+                person_id: person_id,
+                billing_agreement_id: billing_agreement[:billing_agreement_id],
+                active: true
+              })
+
+            Result::Success.new(account)
+          end
+        }
+
+      }
+    end
+
+    def delete_billing_agreement(community_id, person_id)
+      PaypalAccountStore.delete_billing_agreement(
+        {
+          person_id: person_id,
+          community_id: community_id
+        })
+
+      Result::Success.new()
+    end
+
+    ## DELETE /accounts/:community_id/:person_id
+
+    def delete(community_id, person_id=nil)
+      Result::Success.new(PaypalAccountStore.delete(person_id, community_id))
+    end
+
+    ## GET /accounts/:community_id(/:person_id?)
+
+    def get(community_id, person_id=nil)
+      Result::Success.new(PaypalAccountStore.get(person_id, community_id))
+    end
+
+    private
+
+    def with_permission_request_response(callback_url, &block)
```

Comment:
Instead of per request specific helpers payments.rb has a pattern that uses one more general `with_success` helper to do the request. Upside is that the request type and data used is more apparent already at api implementing method. Could we do something similar here? Have a loko at request_wrapper.rb